### PR TITLE
Check for users and roles in the product module

### PR DESCRIPTION
**Pull request by staticf0x**
Relates to #217 (the `errata_tool_product` module part) and #108 

I moved the user fetching code into the `common_errata_tool` shared module so that it can be reused in the `errata_tool_release` module.

### CODE REVIEW

**Comment by ktdreyer** [comment]
Thank you very much for writing this, and especially for writing docstrings and tests!

**Summarizing my thoughts: I appreciate the work you've done, and let's break it down into smaller commits and simplify the implementation.**

Detailed review below.

------

User validation stuff has been a work in progress as I've made different discoveries around API quirks. I'm sorry it's not very straightforward for you to add this feature.

There are two other separate areas of the codebase that also make GET requests to `/api/v1/user`. When we add a third method without refactoring those other two, it's complicated to read the history or clean this up afterwards, so I want to address this before we take this change.

The biggest thing that stands out to me from your PR is that we should split this up into several smaller logical changes to make this all easier to review, like:

- [ ] Resolve #109 and get rid of the hacks for CLOUDWF-8 in `errata_tool_user.get_user()`. (Don't raise `UserNotFoundError` yet.)
- [ ] Move the `errata_tool_user.get_user()` method out of `errata_tool_user.py`, into `common_errata_tool.py`.  (Still do not raise `UserNotFoundError`.)
- [ ] Make `common_errata_tool.get_user()` optionally raise `UserNotFoundError` (add a `strict=True` kwarg) instead of returning `None`.
- [ ] Refactor `common_errata_tool.user_id()` to use `get_user(..., strict=True)`.

I'm going to submit these changes today.

Would you mind rebasing this PR on that?

------

Only one of errata-tool-ansible's resources (ie Products) requires a user role at this point (#216). I don't foresee the Errata Tool developers adding more role restrictions to the server anytime soon, because more ACL management always leads to more manual ops churn work, escalations, etc, so there will always be a natural pressure to keep roles lightweight.

Instead of building a general `InsufficientRoles` class today that can handle any future arbitrary role problems, let's follow [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and write a simple `NoDocsRoleError` class instead. That will make this code a lot easier to read and understand.

-------

There are no unit tests for `validate_docs_reviewer()` method yet, and in fact, it will be difficult to unit test all the code paths. For example, we'd need to add tests for `check_mode` True/False, `ANSIBLE_STRICT_USER_CHECK_MODE` true/false/unset, `default_docs_reviewer` absent/present, etc. This tells me this method is doing too much. It's also a bit difficult to explain in plain English. For example: "Why are we calling a method to validate an account if the user did not even set an account"?

We do not need to validate the docs reviewer user account every time Ansible runs in check mode. We only need to validate the docs reviewer if Ansible would have changed something. In other words, you only need to validate the docs reviewer *after* you've called the main `result = ensure_product()` method and `result['changed']` is True.

Here are some suggestions for reducing or eliminating `validate_docs_reviewer()`:

* An `AnsibleModule.params` will always have the full set of `argument_spec` keys. This means you can always assume the `default_docs_reviewer` key is present. Don't use `params.get()`.

* Do all the validations earlier. You can simply wrap in an `if check_mode and result['changed'] and params['default_docs_reviewer']:` block.

-------

We need to write documentation for `ANSIBLE_STRICT_USER_CHECK_MODE`:
 - What this environment variable does,
 - An example of how a playbook author would turn it on,
 - What an example error message from `ansible-playbook` looks like.

Please add a section to the README, right below the `ERRATA_TOOL_URL` description.

-------

This PR does not entirely close #217 yet, because it adds `ANSIBLE_STRICT_USER_CHECK_MODE` logic to `errata_tool_product` but not `errata_tool_release`. Please edit the description so we're not closing that ticket yet.

------

Other low-priority style-related notes:

- From [the Python 3 docs](https://docs.python.org/3/tutorial/errors.html), "Most exceptions are defined with names that end in "Error", similar to the naming of the standard exceptions." Let's rename all new exceptions to follow this pattern.

- It's great that you've written docstrings. It looks like those follow the Google Python Style guide instead of the Sphinx style that I've used in the rest of the codebase and that Ansible uses. I get that those are harder to read than Google's, though. The reason I do this is so that we can eventually generate HTML documentation on readthedocs.org for these modules, although that's a ways off, and maybe we could use napoleon for that translation.

- Re-ordering `import` statements is nice, did your editor do that automatically? It's a separate logical change that's distracting to me as I review this. In the future, please move those type of import reordering changes to a separate commit or even a separate PR.

- Similarly, please don't switch to an entirely new import style with `from ansible.module_utils.common_errata_tool import (giant tuple)...` in this change. When we cram that list into a tuple, especially with multiple elements on single lines, with no dangling commas, it makes it [harder to edit that code over time](https://medium.com/@nikgraf/why-you-should-enforce-dangling-commas-for-multiline-statements-d034c98e36f8).

**Comment by ktdreyer** [comment]
This is fantastic, thank you!

Today I've pushed test coverage for the existing `main()` method in https://github.com/ktdreyer/errata-tool-ansible/pull/239

You can rebase onto that, and then add more tests to the `TestMain` class that will use `monkeypatch.setenv('ANSIBLE_STRICT_USER_CHECK_MODE', ...)`

**Comment by meesha7** [comment]
> This is fantastic, thank you!
> 
> Today I've pushed test coverage for the existing `main()` method in #239
> 
> You can rebase onto that, and then add more tests to the `TestMain` class that will use `monkeypatch.setenv('ANSIBLE_STRICT_USER_CHECK_MODE', ...)`

I added the tests, but py27 keeps failing (see #240)
